src/Compilers/CSharp/Portable/Symbols/Source/SourceLocalSymbol.cs

```diff
@@ -580,5 +598,58 @@ protected override TypeSymbol InferTypeOfVarVariable(DiagnosticBag diagnostics)
                 }
             }
         }
+
+        /// <summary>
+        /// Symbol for a deconstruction local that might require type inference.
+        /// For instance, local `x` in `var(x, y) = ...` or `(var x, int y) = ...`.
+        /// </summary>
+        private class PossiblyImplicitlyTypedDeconstructionLocalSymbol : SourceLocalSymbol
+        {
+            public PossiblyImplicitlyTypedDeconstructionLocalSymbol(
+                Symbol containingSymbol,
+                Binder binder,
+                TypeSyntax typeSyntax,
+                SyntaxToken identifierToken,
+                LocalDeclarationKind declarationKind)
+            : base(containingSymbol, binder, RefKind.None, typeSyntax, identifierToken, declarationKind)
+            {
+            }
+
+            protected override TypeSymbol InferTypeOfVarVariable(DiagnosticBag diagnostics)
+            {
+                // Try binding immediately enclosing deconstruction-declaration (the top-level VariableDeclaration), this should force the inference.
+
+                var topLevelVariableDeclaration = (VariableDeclarationSyntax)IdentifierToken.
+                                                                    Parent. // VariableDeclaratorSyntax
+                                                                    Parent; // VariableDeclarationSyntax
+                while (true)
+                {
+                    var parent = topLevelVariableDeclaration.Parent;
+                    if (parent.Kind() != SyntaxKind.VariableDeconstructionDeclarator) { break; }
+
+                    var grandParent = parent.Parent;
+                    if (grandParent.Kind() != SyntaxKind.VariableDeclaration) { break; }
+
+                    topLevelVariableDeclaration = (VariableDeclarationSyntax)grandParent;
+                }
+
+                Debug.Assert(topLevelVariableDeclaration.Deconstruction != null && topLevelVariableDeclaration.Deconstruction.Value != null);
+
+                var statement = topLevelVariableDeclaration.Parent;
+                TypeSymbol result;
+                switch (statement.Kind())
+                {
+                    case SyntaxKind.LocalDeclarationStatement:
+                        this.binder.BindDeconstructionDeclarationStatementParts((LocalDeclarationStatementSyntax)statement, diagnostics);
```

Comment:
Opened [https://github.com/dotnet/roslyn/issues/12266](https://github.com/dotnet/roslyn/issues/12266)

---
In reply to: [68833592](https://github.com/dotnet/roslyn/pull/12224#discussion_r68833592) [](ancestors = 68833592,68823222)